### ISSUE TITLE
Implement Asynchronous HTTP Client requests

### ISSUE THREAD

**Issue by LucasMull** [cee-studio/orca#573]
This one is certainly not a priority but I will write for future reference. 

Currently every request is done synchronously, which means that if we have 5 requests waiting on a queue they can only be performed one after the other, blocking the entire thread along with it. We could make this work by providing a optional asynchronous state, which performs its requests on a thread whose sole responsibility is enabling sockets for read/write via poll() (curl_multi_perform() is basically a wrapper around that). This works, but we will need to introduce three new functions (which are very familiar and may trigger PTSD)
```c
struct data_s data;
struct user_agent_s async_ua;

async_begin(ua, &async_ua); // create a unique asynchronous branch to ua

async_await(&async_ua, (void*)&data, callback1);
async_await(&async_ua, (void*)&data, callback2);

// any function receiving async_ua as parameter will be dealt with asynchronously
ua_run(
  &async_ua,
  &resp_handle,
  &req_body,
  HTTP_GET,
  "/endpoint");

ua_run(
  &async_ua,
  &resp_handle,
  &req_body,
  HTTP_POST,
  "/endpoint");
// ua_run requests will run in parallel
async_end(&async_ua); // async_end waits until every request is completed, then run pending await callbacks
```
Anything between async_begin() and async_end() should use async_await() in order to make sure that particular code will run only after the response has been fully received. ua_run will be able to detect whether the request should be asynchronous by its received user_agent

**Comment by mlite** [comment]
We need more data (use cases) to justify this.

**Comment by LucasMull** [comment]
This is unecessary, and much harder to implement ratelimiting with.
